GoogleForCreators/web-stories-wp
# Taxonomy: Autocomplete Options for Flat Terms

## Context

Adds a dropdown menu ala Gutenberg's of available terms that start with the value typed. 

## Summary



## Relevant Technical Choices

- autocomplete suggestions stored as `flatSearchResults` based on `addSearchResultsToCache` - just the response isolated so it's only things that match the current value. Implementation taken from #9211 
- dropdown placement was ripped from the dropdown component (thanks @BrookeGraham ) - this might be a cool thing to extract eventually. 
- Using state to decide when the autocomplete menu's focused. So when arrow up/down (depending on direction menu opened) it gets set to focused, when focus returns to parent from menu it's set to false again and also set to false when a value is submitted. 
- Added `,` to the keys since that's what the key is for `Comma` - left in comma because i'm curious if this was for a browser edge case - haven't looked into it yet. 
- Update `SUBMIT_VALUE` action to take a value optionally so that options from dropdown menu can be treated the same as values added via typing. 

## To-do

- [ ] Fix the tests that are failing
- [x] Discuss with Max what's up with the offset reset and update accordingly.
- [x] Test RTL 
- [x] Test roles 

## User-facing changes

(Behind feature flag) 
<img width="315" alt="Screen Shot 2021-09-30 at 3 28 55 PM" src="https://user-images.githubusercontent.com/10720454/135539368-acfe6cb3-736f-45fe-a5ba-267e192eba70.png">



## Testing Instructions

Go to the tags section after enabling the taxonomy experiment and add some tags. You should be able to still add new tags (if you're admin) and anytime there's tags that match what your typed value starts with and you've typed more than 2 characters you'll get a menu with those options. You can select options from there with your keyboard or your cursor. It should mimic Gutenberg's.


## Reviews

### Does this PR have a security-related impact?

No

### Does this PR change what data or activity we track or use?

No

### Does this PR have a legal-related impact?

No

## Checklist




---



Fixes #9034
